pydicom/fileset.py

```diff
@@ -0,0 +1,2720 @@
+            number of elements are available).
+
+        Returns
+        -------
+        list of object
+            A list of value(s) for the element available in the instances.
+        """
+        has_element = False
+        results = []
+        instances = instances or iter(self)
+        for instance in instances:
+            if load:
+                instance = instance.load()
+
+            if element not in instance:
+                continue
+
+            has_element = True
+            val = instance[element].value
+            # Not very efficient, but we can't use set
+            if val not in results:
+                results.append(val)
+
+        if not load and not has_element:
+            warnings.warn(
+                "None of the records in the DICOMDIR dataset contain "
+                "the query element, consider using the 'load' parameter "
+                "to expand the search to the corresponding SOP instances"
+            )
+
+        return results
+
+    @property
+    def ID(self) -> Union[str, None]:
+        """Return the *File-set ID* (if available) or ``None``."""
+        return self._id
+
+    @ID.setter
+    def ID(self, val: Union[str, None]) -> None:
+        """Set the File-set ID.
+
+        Parameters
+        ----------
+        val : str or None
+            The value to use for the DICOMDIR's (0004,1130) *File-set ID*.
+
+        Raises
+        ------
+        ValueError
+            If `val` is greater than 16 characters long.
+        """
+        if val == self._id:
+            return
+
+        if val is None or 0 <= len(val) <= 16:
+            self._id = val
+            if self._ds:
+                self._ds.FileSetID = val
+            self._stage['^'] = True
+        else:
+            raise ValueError(
+                "The maximum length of the 'File-set ID' is 16 characters"
+            )
+
+    @property
+    def is_staged(self) -> bool:
+        """Return ``True`` if the File-set is new or has changes staged."""
+        is_staged = [
+            self._stage['+'],
+            self._stage['-'],
+            self._stage['~'],
+            self._stage['^'],
+        ]
+        return any(is_staged)
+
+    def __iter__(self) -> FileInstance:
+        """Yield :class:`~pydicom.fileset.FileInstance` from the File-set."""
+        yield from self._instances
+
+    def __len__(self) -> int:
+        """Return the number of instances in the File-set."""
+        return len(self._instances)
+
+    def load(
+        self,
+        ds_or_path: Union[Dataset, str, PathLike],
+        include_orphans: bool = True,
+        raise_orphans: bool = False,
+    ) -> None:
+        """Load an existing File-set.
+
+        Existing File-sets that do not use the same directory structure as
+        *pydicom* will be staged to be moved to a new structure. This is
+        because the DICOM Standard attaches no semantics to *how* the files
```

Comment:
Staging: if you're changing the file-set, rather than immediately write those changes to the file then they get "staged", similarly to how git works. Once you've staged your changes you call `FileSet.write()` (git's commit) to apply them to the file system.

The directory structure thing is the DICOM Standard doesn't say *how* the files in a file-set have to be laid out. They can all be in the same directory, they can each have their own directory, etc, etc. Because it's simply not possible to modify an existing directory structure without knowing its semantics, the simplest way to manage any modification (adding/removing instances) is to convert the entire directory structure to the one used by `FileSet`.

So for a file with 4 records (PATIENT, STUDY, SERIES, IMAGE) we end up with:
```
PT000000/ST000000/SE000000/IM000000
```
Where `IM000000` is the file.

Reply:
I can see that is reasonable because, as you say, it would be difficult to try to determine the user's structure.  However, if only removing then could stay with the same structure.

I see that it says the files are moved... it appears that only happens when the write method is called, is that correct?  Moving files seems a touchy subject (potentially slow if thousands of them, confusing to user if not aware that that is what will happen, maybe the code doesn't work quite the same on running a second time).  I don't know what I would suggest - I thought of maybe a flag to move (default False), to force the user to indicate they are okay with that, but I don't love that either...  Is it possible at least for someone to just create a DICOMDIR with an existing directory structure?  I think the expected behavior in that case would be to leave the file locations alone. (Likewise for the remove-only case I already mentioned).  I thought I saw something in the code that parsed existing directory by just staging for add, so it would seem they all follow the same path and maybe it is not easy to special-case like that with the current code structure?

Reply:
> I see that it says the files are moved... it appears that only happens when the write method is called, is that correct?

Yup. No (permanent) changes to the file system occur unless `write()` is called.

If they were only removing then it could work, but it'd have to be an explicit flag for `write()` because the behaviour would be different. It'd be similar to the current `dicomdir_only` flag, which keeps the existing structure when only a DICOMDIR update is needed. Another choice was to keep the existing files where they are and add the new structure alongside it, but I figured managing the File IDs and avoiding collisions would be more difficult.

And from a usability point of view I've found that I prefer the PT/ST/SE/IM structure when I have to deal with patient DVDs because I can just copy + paste the correct scan without needing third party tools (side note: whoever thought that putting all the data into a single directory needs to reconsider their life choices).

Creating a file-set from an existing directory structure by just adding a DICOMDIR is definitely something that could be added fairly easily (I actually started out from that end - it's why `is_conformant_file_id()` exists), but I've tried to keep the current PR to the minimum functionality required. I think it's better to have *something*, even if its not perfect, that can be improved upon. 

Speaking of which, another nice enhancement would be to add a method to create a README listing how the directory structure corresponds to the patient/study/series.

Reply:
> I think it's better to have _something_, even if its not perfect, that can be improved upon.

Yes, good point. I'll have another look at the `dicomdir_only` flag, and I still have to go back for a second pass for all of it.  It may take a couple of days given the size of the code added.

> And from a usability point of view I've found that I prefer the PT/ST/SE/IM structure when I have to deal with patient DVDs because I can just copy + paste the correct scan without needing third party tools (side note: whoever thought that putting all the data into a single directory was a good idea needs to reconsider their life choices).

lol... yep.  But with PT/ST/SE/IM, it would be mostly self-documenting and maybe the DICOMDIR would not really be necessary  :wink:

Reply:
> It may take a couple of days given the size of the code added

I still have the file-set tutorial to finish, anyway, so no rush

Reply:
I've added an `use_existing` flag that also replaces `dicomdir_only`.